CMakeLists.txt

```diff
@@ -64,9 +64,15 @@ target_link_directories(notcurses
 )
 endif()
 
+if(CMAKE_BUILD_TYPE STREQUAL Debug)
+  set(DEBUG_OPTIONS -g -O0)
+else()
```

Comment:
Hrmmm, does it not handle this for us already? Whenever I've done a `cmake .. -DCMAKE_BUILD_TYPE=Debug` (which I definitely do when valgrinding), I get full callgraphs and line numbers and all that good stuff.

If CMake can't do this kind of basic thing, I'm really disappointed in it as any kind of "abstracting" build tool :/.

Reply:
```
[grimes](0) $ cmake ../ -DCMAKE_BUILD_TYPE=Debug > /dev/null
[grimes](0) $ touch ../src/view/view.cpp 
[grimes](0) $ make VERBOSE=1 src/view/view.o
make -f CMakeFiles/notcurses-view.dir/build.make CMakeFiles/notcurses-view.dir/src/view/view.cpp.o
make[1]: Entering directory '/home/dank/src/dankamongmen/notcurses/build'
Building CXX object CMakeFiles/notcurses-view.dir/src/view/view.cpp.o
/usr/sbin/c++  -DFORTIFY_SOURCE=2 -D_XOPEN_SOURCE -I/home/dank/src/dankamongmen/notcurses/include -I/home/dank/src/dankamongmen/notcurses/build/include -I/home/dank/src/dankamongmen/notcurses  -g -fvisibility=hidden   -Wall -Wextra -W -Wshadow -std=c++17 -o CMakeFiles/notcurses-view.dir/src/view/view.cpp.o -c /home/dank/src/dankamongmen/notcurses/src/view/view.cpp
make[1]: Leaving directory '/home/dank/src/dankamongmen/notcurses/build'
[grimes](0) $ 
```

so Debug already seems to (at least with my CMake 3.15) add `-g`, and supply no `-O`, which is equivalent to explicitly supplying `-O0` AFAIK. If I then build `Release`-style:

```
 [grimes](0) $ make VERBOSE=1 src/view/view.o
make -f CMakeFiles/notcurses-view.dir/build.make CMakeFiles/notcurses-view.dir/src/view/view.cpp.o
make[1]: Entering directory '/home/dank/src/dankamongmen/notcurses/build'
Building CXX object CMakeFiles/notcurses-view.dir/src/view/view.cpp.o
/usr/sbin/c++  -DFORTIFY_SOURCE=2 -D_XOPEN_SOURCE -I/home/dank/src/dankamongmen/notcurses/include -I/home/dank/src/dankamongmen/notcurses/build/include -I/home/dank/src/dankamongmen/notcurses  -O3 -DNDEBUG -fvisibility=hidden   -Wall -Wextra -W -Wshadow -std=c++17 -o CMakeFiles/notcurses-view.dir/src/view/view.cpp.o -c /home/dank/src/dankamongmen/notcurses/src/view/view.cpp
make[1]: Leaving directory '/home/dank/src/dankamongmen/notcurses/build'
[grimes](0) $ 
```

I get -O3, which though a bit much during peacetime isn't actually offensive. So I'm not sure I get the point of this addition?

Reply:
I guess it depends on the OS. With the defaults (on Ubuntu 19.10, GCC 9) I get:

```
[1/100] /usr/lib/ccache/cc -DFORTIFY_SOURCE=2 -DSOURCE_DEFAULT -D_GNU_SOURCE -D_XOPEN_SOURCE -I../include -I../ -fvisibility=hidden   -std=gnu11 -MD -MT CMakeFiles/rgbbg.dir/src/poc/rgbbg.c.o -MF CMakeFiles/rgbbg.dir/src/poc/rgbbg.c.o.d -o CMakeFiles/rgbbg.dir/src/poc/rgbbg.c.o   -c ../src/poc/rgbbg.c
[2/100] /usr/lib/ccache/c++  -DFORTIFY_SOURCE=2 -DSOURCE_DEFAULT -D_GNU_SOURCE -D_XOPEN_SOURCE -I../include -Iinclude -fvisibility=hidden   -std=c++17 -MD -MT CMakeFiles/notcurses++-static.dir/src/libcpp/NotCurses.cc.o -MF CMakeFiles/notcurses++-static.dir/src/libcpp/NotCurses.cc.o.d -o CMakeFiles/notcurses++-static.dir/src/libcpp/NotCurses.cc.o -c ../src/libcpp/NotCurses.cc
```

And `notcurses-view` binary looks as follows:

```
$ ls -l notcurses-view 
-rwxrwxr-x 1 grendel grendel 65056 Jan  9 22:28 notcurses-view

$ file notcurses-view 
notcurses-view: ELF 64-bit LSB pie executable, x86-64, version 1 (GNU/Linux), dynamically linked, interpreter /lib64/ld-linux-x86-64.so.2, BuildID[sha1]=dfb6287b161d5eb64ec600a04fbed628bc3c25f7, for GNU/Linux 3.2.0, not stripped
```

while with the above chunk in `CMakeLists.txt`, I get:

```
$ ls -l notcurses-view 
-rwxrwxr-x 1 grendel grendel 375456 Jan  9 22:33 notcurses-view

$ file notcurses-view 
notcurses-view: ELF 64-bit LSB pie executable, x86-64, version 1 (GNU/Linux), dynamically linked, interpreter /lib64/ld-linux-x86-64.so.2, BuildID[sha1]=ecaa1200179f3a74c4d4dc6c1e8f5c074ac3be26, for GNU/Linux 3.2.0, with debug_info, not stripped
```

Reply:
Hrmmm, I'm guessing this comes down to CMake version? I'm running 3.15.4 locally. I'm loathe to bulk up our CMakeLists.txt with stuff that CMake ought really be handling for us, but you do have evidence here...

Reply:
I've seen this behavior from 3.10 through 3.16 at least. I agree it should be done by cmake, but since the reality disagrees... :)

Reply:
yep yep, it's all good.